File: src/conventions/core/bundle/bundle-types.ts
import type { DealConstraints, Deal, Auction } from "../../../engine/types";
import type { Seat } from "../../../engine/types";
import type { AlternativeGroup, SurfaceGroup } from "../../../core/contracts/teaching-grading";
import type { TeachingRelation } from "../../../core/contracts/teaching-projection";
import type { SystemProfile } from "../../../core/contracts/agreement-module";
import type { ConventionConfig, ConventionTeaching } from "../../../core/contracts/convention";
import type { ConventionCategory } from "../../../core/contracts/convention";
import type { ConventionModule } from "../convention-module";

// ── Authored input ──────────────────────────────────────────────────

/**
 * What convention authors hand-write when defining a bundle.
 *
 * Deal constraints, off-convention constraints, default auction, and
 * allowed dealers are DERIVED from capabilities + R1 surface analysis
 * by `deriveBundleDealConstraints()` in `system-registry.ts`. Authors
 * declare `declaredCapabilities` and the derivation handles the rest.
 */
export interface BundleInput {
  readonly id: string;
  readonly name: string;
  readonly memberIds: readonly string[];
  /** If true, hidden from UI picker (e.g., parity testing bundle). */
  readonly internal?: boolean;
  /** Optional system profile for profile-based module activation. */
  readonly systemProfile?: SystemProfile;
  /** Capabilities to inject into profile-based activation. Only capabilities
   *  declared here are provided — bundles without this field get no capabilities.
   *  Also used by deal constraint derivation to determine opener constraints. */
  readonly declaredCapabilities?: Readonly<Record<string, string>>;
  /** Convention category for UI grouping. */
  readonly category: ConventionCategory;
  /** Human-readable description for UI display. */
  readonly description: string;
  /** Convention-level teaching metadata for learning UI. */
  readonly teaching?: ConventionTeaching;
  // ruleModules removed — modules are resolved by buildBundle() from memberIds via module-registry.
}

// ── Derived teaching content ────────────────────────────────────────

/**
 * Teaching/grading metadata derived from module content at bundle build time.
 * Computed by `buildBundle()`, NOT hand-authored.
 *
 * Stored alongside the bundle and passed to evaluation consumers for
 * bid grading (acceptable alternatives, near-miss detection).
 */
export interface DerivedTeachingContent {
  /** @derived From ALTERNATIVES teachingTags on surfaces. */
  readonly acceptableAlternatives: readonly AlternativeGroup[];
  /** @derived From teachingTags + rule module structure. */
  readonly surfaceGroups: readonly SurfaceGroup[];
  /** @derived From relation teachingTags (same-family, stronger-than, etc.) on surfaces. */
  readonly relations: readonly TeachingRelation[];
}

// ── Full computed bundle ────────────────────────────────────────────

/**
 * Complete convention bundle: authored input + derived constraints + derived teaching.
 *
 * Produced by `resolveBundle()`. Deal constraints are derived from capabilities
 * and R1 surface analysis — not hand-authored.
 */
export interface ConventionBundle extends BundleInput {
  /** Resolved convention modules (assembled by resolveBundle from memberIds). */
  readonly modules: readonly ConventionModule[];
  /** Derived teaching/grading metadata (acceptableAlternatives, surfaceGroups, relations). */
  readonly derivedTeaching: DerivedTeachingContent;
  /** @derived Deal constraints from capability archetype + R1 surface analysis. */
  readonly dealConstraints: DealConstraints;
  /** @derived Off-convention constraints via complement negation. */
  readonly offConventionConstraints?: DealConstraints;
  /** @derived Default auction from capability archetype. */
  readonly defaultAuction?: (seat: Seat, deal?: Deal) => Auction | undefined;
  /** @derived Allowed dealers from capability archetype. */
  readonly allowedDealers?: readonly Seat[];
}

/**
 * Creates a ConventionConfig from a ConventionBundle.
 * All metadata is derived from the bundle — no overrides needed.
 */
export function createConventionConfigFromBundle(
  bundle: ConventionBundle,
): ConventionConfig {
  return {
    id: bundle.id,
    name: bundle.name,
    description: bundle.description,
    category: bundle.category,
    dealConstraints: bundle.dealConstraints,
    offConventionConstraints: bundle.offConventionConstraints,
    defaultAuction: bundle.defaultAuction,
    internal: bundle.internal,
    teaching: bundle.teaching,
    allowedDealers: bundle.allowedDealers,
  };
}

/**
 * Re-derive a ConventionConfig's deal constraints for a different SystemConfig.
 *
 * With deal constraint derivation, constraints are always system-parameterized
 * via the capability archetype. The bundle's constraints are already derived
 * for the active system — re-resolve the bundle for a different system.
 *
 * @deprecated Prefer re-resolving the bundle via `resolveBundle(input, newSys)`.
 */
export function resolveConventionForSystem(
  config: ConventionConfig,
  _bundle: ConventionBundle | undefined,
  _sys: unknown,
): ConventionConfig {
  // Constraints are now always derived at bundle resolution time.
  // No factory-based re-derivation needed — the bundle already has
  // the correct constraints for its system.
  return config;
}

File: crates/bridge-conventions/src/types/bundle_types.rs
//! Convention bundle types.
//!
//! Mirrors TS types from `conventions/core/bundle/bundle-types.ts`
//! and `conventions/core/convention-types.ts`.

use bridge_engine::types::{DealConstraints, Seat};
use serde::{Deserialize, Serialize};

use super::agreement::SystemProfile;
use super::module_types::ConventionModule;
use super::teaching::SurfaceGroup;

/// Convention category for UI grouping.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ConventionCategory {
    Asking,
    Defensive,
    Constructive,
    Competitive,
}

/// Convention-level teaching metadata.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConventionTeaching {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub purpose: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub when_to_use: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub when_not_to_use: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tradeoff: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub principle: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roles: Option<String>,
}

/// What convention authors hand-write when defining a bundle.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BundleInput {
    pub id: String,
    pub name: String,
    pub member_ids: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub internal: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_profile: Option<SystemProfile>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub declared_capabilities: Option<std::collections::HashMap<String, String>>,
    pub category: ConventionCategory,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub teaching: Option<ConventionTeaching>,
}

/// Teaching/grading metadata derived from module content at bundle build time.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DerivedTeachingContent {
    pub surface_groups: Vec<SurfaceGroup>,
}

/// Complete convention bundle: authored input + derived constraints + derived teaching.
/// Function fields (defaultAuction) are omitted — data only.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConventionBundle {
    // BundleInput fields
    pub id: String,
    pub name: String,
    pub member_ids: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub internal: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_profile: Option<SystemProfile>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub declared_capabilities: Option<std::collections::HashMap<String, String>>,
    pub category: ConventionCategory,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub teaching: Option<ConventionTeaching>,

    // Derived fields — populated at runtime from module registry, not in fixture JSON
    #[serde(default)]
    pub modules: Vec<ConventionModule>,
    pub derived_teaching: DerivedTeachingContent,
    pub deal_constraints: DealConstraints,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub off_convention_constraints: Option<DealConstraints>,
    // defaultAuction omitted — function field
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allowed_dealers: Option<Vec<Seat>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supports_role_selection: Option<bool>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn convention_category_serde() {
        assert_eq!(
            serde_json::to_string(&ConventionCategory::Asking).unwrap(),
            "\"Asking\""
        );
        assert_eq!(
            serde_json::to_string(&ConventionCategory::Defensive).unwrap(),
            "\"Defensive\""
        );
    }

    #[test]
    fn convention_teaching_roundtrip() {
        let teaching = ConventionTeaching {
            purpose: Some("Find a 4-4 major fit".to_string()),
            when_to_use: Some("After partner opens 1NT".to_string()),
            when_not_to_use: Some(vec!["With 4-3-3-3 shape".to_string()]),
            tradeoff: Some("Can't play in 2C".to_string()),
            principle: None,
            roles: None,
        };
        let json = serde_json::to_string(&teaching).unwrap();
        let back: ConventionTeaching = serde_json::from_str(&json).unwrap();
        assert_eq!(back, teaching);
    }
}
